### ISSUE TITLE
AutomationPeer.GetRuntimeId assumes Object.GetHashCode() is unique

### ISSUE THREAD

**Issue by madewokherd** [dotnet/wpf#8096]
### Description

AutomationPeer.GetRuntimeId returns an array containing the process id and the result of GetHashCode for the AutomationPeer object. Since it doesn't use UiaAppendRuntimeId (which technically is relying on undocumented behavior already), those values must uniquely identify the UIAutomation element. None of the subclasses override GetHashCode or GetRuntimeId, so Object.GetHashCode is used. The [MSDN documentation for Object.GetHashCode](https://learn.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-8.0#system-object-gethashcode) says this:

> Two objects that are equal return hash codes that are equal. However, the reverse is not true: equal hash codes do not imply object equality, because different (unequal) objects can have identical hash codes. Furthermore, .NET does not guarantee the default implementation of the [GetHashCode](https://learn.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-8.0) method, and the value this method returns may differ between .NET implementations, such as different versions of .NET Framework and .NET Core, and platforms, such as 32-bit and 64-bit platforms. For these reasons, do not use the default implementation of this method as a unique object identifier for hashing purposes.

### Reproduction Steps

Object hash code collisions are likely rare, so this would be difficult to reproduce.

### Expected behavior

GetRuntimeId should return a value that is guaranteed to uniquely identify the element.

### Actual behavior

GetRuntimeId may not uniquely identify the element because it depends on the result of Object.GetHashCode to uniquely identify the object.

### Impact

May cause rare and difficult-to-reproduce errors in Assistive Technologies when they interact with WPF applications. UI Automation internally relies on GetRuntimeId to be unique across the desktop (Wine tests have examples of this), and Assistive Technologies may also make this assumption.

### Other information

My suggestion would be to add a static field with a monotonic counter used to generate runtime id's in AutomationPeer's constructor.

**Comment by anjali-wpf** [comment]
@madewokherd - The purpose of GetHashCode is not to generate a unique identifier for an object, but to implement data structures that are based on hash tables, such as Dictionary<K, V> or HashSet<T>.

A hash function is required to ensure that if `x == y`, then `x.GetHashCode() == y.GetHashCode()`, but the converse is not true: Two distinct objects can have the same hash code. This situation is called a hash collision.

Hash table structures will still work if there are collisions, but they run slower because your program has to spend time disambiguiting which of the collided object you're searching for. So, a good hash function will strive to minimize collisions. (Note that it's a mathematical impossibility to completely avoid collisions if there are more than 232 possible values for a class, because of the pigeonhole principle.)

`GetHashCode()` is not what you should be using to hash a string, almost 100% of the time. I recommend that you use an actual hash algorithm, like `SHA-1`:

```
using(System.Security.Cryptography.SHA1Managed hp = new System.Security.Cryptography.SHA1Managed()) {
    // Use hp.ComputeHash(System.Text.Encoding.ASCII (or Unicode, UTF8, UTF16, or UTF32 or something...).GetBytes(theString) to compute the hash code.
}

```
For something a little bit faster, there's also `SHA1Cng`, which is significantly faster than `SHA1Managed`.

**Comment by madewokherd** [comment]
> @madewokherd - The purpose of GetHashCode is not to generate a unique identifier for an object, but to implement data structures that are based on hash tables, such as Dictionary<K, V> or HashSet.

Yes, that's exactly why GetHashCode shouldn't be used here. UIAutomation has to rely only on the key it gets for comparison, and that key only has the process id and hash code for the object.
